[Resources] Add resources clause and typed buffers (#65)
hlsl-tc57/tc57: spec/resources.tex

\Ch{Resource Types}{Resources}

> \p Resources are intangible types representing memory with an external interface.

Reviewer: I think it would be less confusing to refer to the types as "Resource types" and instances of them as "resources"

While I suppose there is some memory associated with a SamplerState object, I wouldn't consider it to represent memory in the same way as a texture or buffer. Maybe something like this?
```suggestion
\p Resource types are intangible types representing memory or configuration with an external interface.
```

Reviewer: Yes, I'm aware I wrote this

\p These take the form of Typed Buffers, Raw Buffers, Textures, Constant Buffers
and Samplers.

\p Typed Buffer, Raw Buffer, and Texture resources may be read-only or
read-write as denoted by the type of the resource, where the names of read-write

> resources are prefixed with \texttt{RW}.

Reviewer: ```suggestion
resource types are prefixed with \texttt{RW}.
```

\p Global resource declarations are \textit{bound} to device memory managed by
an application through a runtime implementation. A resource declaration that is
not bound is said to be \textit{unbound}. Operations on unbound resources are
evaluated as no-ops.

\p An implementation may impose restrictions on how local resource objects may
be used (\ref{AnnexB.LocalResources}).

\Sec{Typed Buffers}{Resources.TypedBuffers}

> \p The typed buffer class template represents a one-dimensional resource

Reviewer: It's capitalized above
```suggestion
\p The Typed Buffer class template represents a one-dimensional resource
```

Author: I don't think these are proper nouns. They're capitalized in the title, but not the text.

Reviewer 2: I think Greg is referring to their capitalisation not in the section title, but the previous section. https://github.com/hlsl-tc57/tc57/pull/65/changes#diff-f43f5ddb995c0acc6ae11cd5c41b19f95a84457e4416d8fe0405b3926e7df1c7R5

Author: Ah, I've updated the earlier section.

The total size of a single contained element must be less than 16 bytes.

\p Typed buffers perform format conversions on load such that the underlying
data gets converted to the destination type in an implementation-defined way.

\begin{HLSL}
template <typename T = float4> requires hlsl::is_vector<T>::value || hlsl::is_arithmetic<T>::value
class Buffer {
  Buffer();
  Buffer(const Buffer &buf);
  Buffer &operator=(Buffer &buf);
  Buffer(const __DynamicResourceHandle &handle);

  void GetDimensions(out uint width) const;

  // Element access.
  T operator[](int index) const;
  T Load(int index) const;
  T Load(int index, out uint status) const;
};

template <typename T> requires hlsl::is_vector<T>::value || hlsl::is_arithmetic<T>::value
class RWBuffer {
  RWBuffer();
  RWBuffer(RWBuffer buf);
  RWBuffer &operator=(RWBuffer &buf);
  Buffer(const __DynamicResourceHandle &handle);

  void GetDimensions(out uint width) const;

  // Element access.
  T operator[](int index) const;
  device T &operator[](int n) const;
  T Load(int index) const;
  T Load(int index, out uint status) const;
};
\end{HLSL}

\Sub{Constructors}{Resources.TypedBuffers.Ctrs}

> 

Reviewer: I see you removed the language about default template parameters. Removing that support is a breaking change

Author: The language about default template parameters doesn't belong here. It isn't unique to resource type it applies to _all_ template types and should be part of the templates clause.

Reviewer: It doesn't even apply to all resource types. Multisampled textures require template arguments: https://godbolt.org/z/qz1MGha58

Author: Whether or not a given resource type has a default argument is explicit in the definition of the type. See the above code:

```
template <typename T = float4> requires hlsl::is_vector<T>::value || hlsl::is_arithmetic<T>::value
class Buffer {
...
};

template <typename T> requires hlsl::is_vector<T>::value || hlsl::is_arithmetic<T>::value
class RWBuffer {
...
};
```

`Buffer` has a default parameter, `RWBuffer` does not. Is there a reason to also explicitly spell this out?

The fact that the default parameter allows the type to be spelled without specifying the type or empty template arguments isn't unique to these templates, that's just HLSL's template behavior.

Reviewer: As a result of the committee's decision that code blocks are non-normative, this isn't going to be sufficient. We'll need some language explaining this as well.

Author: Please provide specifics as to what you believe needs to change. The language I removed described how default template arguments behave, which is not something this clause should be addressing. The language [here](https://github.com/hlsl-tc57/tc57/pull/65/changes#diff-f43f5ddb995c0acc6ae11cd5c41b19f95a84457e4416d8fe0405b3926e7df1c7R33) describes the constraints on the template argument, and the (newly added) line [here](https://github.com/hlsl-tc57/tc57/pull/65/changes#diff-f43f5ddb995c0acc6ae11cd5c41b19f95a84457e4416d8fe0405b3926e7df1c7R40) describes the default argument for the `Buffer` type.

Is there any further change required?

\p When defined at local scope, typed buffers represent local references that
can be associated with global buffers when assigned. Use of an uninitialized
typed buffer is undefined.

\begin{HLSL}
Buffer<int> grobuf;
RWBuffer<int> grwbuf;
void main() {
  Buffer<int> lrobuf = grobuf;
  RWBuffer<int> lrwbuf = grwbuf;
}
\end{HLSL}

\p Buffers cannot be implicitly nor explicitly cast from one type to another.

\p Buffers may be created from dynamic resources. The
\texttt{\_\_DynamicResourceHandle} type in the examples above is a placeholder
for the implementation-specific type returned by subscript indexing into the
\texttt{ResourceDescriptorHeap} (\ref{Resources.Binding.Dynamic}).

\Sub{Dimensions}{Resources.TypedBuffers.Dims}

\p The size of a typed buffer can be retrieved using the \texttt{GetDimensions}
method.

\begin{HLSL}
void GetDimensions(out uint width) const;
\end{HLSL}

\p This returns the full length, in elements of the buffer through the
\texttt{width} output parameter.

\Sub{Element Access}{Resources.TypedBuffers.Access}

\p The contents of typed buffers can be retrieved using the \texttt{Load} methods
or the subscript operator.

\begin{HLSL}

> T Load(in int index) const;

Reviewer: As you removed the `in` qualifiers above relying on the reader to assume implicit behavior, it seems consistent to remove it here too
```suggestion
T Load(int index) const;
```

T operator[](int index) const;
\end{HLSL}

\p These each return the element of the buffer's type at the given index
\texttt{index}.

\p An additional \texttt{Load} method returns the indicated element just as the
other, but also takes a second \texttt{status} parameter that returns
information about the accessed resource.

\begin{HLSL}

> T Load(in int index, out uint status) const;

Reviewer: as above
```suggestion
T Load(int index, out uint status) const;
```

\end{HLSL}

\p The \texttt{status} parameter returns an indication of whether all of the
retrieved value came from fully mapped parts of the resource. This parameter
must be passed to the built-in function \texttt{CheckAccessFullyMapped}
(\ref{Resources.Mapcheck}) in order to interpret it.

\p Writable buffers have an additional subscript operator that allows assignment
to an element of the buffer. It behaves as if it returns a reference to that

> element as stored in device memory.

Reviewer: Is this really so different from an mmap concept? I don't think that requires any special specification syntax to indicate that you might be writing into something that isn't actually normal memory. As far as the language is concerned, what difference does it make?

Author: There is actually a semantic problem that is subtle which I'm trying to address here. The real semantics of the subscript operator for a `RWBuffer<T>` isn't that it actually returns a reference to device memory.

The more correct way to model it would be that the subscript operator returns a proxy object, that when written to generates a call that may convert or otherwise mutate the data before the write occurs. This gets _really_ gnarly when you start factoring in the atomic functions, and I haven't really come up with a better way to spec it than this.

\begin{HLSL}
device T &operator[](int index);
\end{HLSL}

\begin{note}
  Accurately modeling the memory operations for writable typed buffers would
  require a proxy-object interface since the address written to isn't actually
  the device memory since it may be converted or otherwise normalized by the
  implementation.
\end{note}

\p Writes to typed buffer elements overwrite the entire element, even if only
part of the data is written. The behavior of partial writes is

> implementation-defined.

Reviewer: I think this should be undefined

Author: What's the argument for undefined rather than implementation-defined? The later lets an implementation do whatever it wants as long as it documents it.

Reviewer: I am saying that it should be undefined instead of implementation-defined. 

Is there a platform you have in mind here or do you just want to keep options open? Requiring treatment of each element as a unit that requires full updating is pretty close to the definition of a typed buffer. I think failing to do that should be treated as invalid.

Author: I believe this should be implementation-defined because each implementation of HLSL should specify this behavior, which allows each implementation to have a different behavior as long as it is well-defined.

Why should it be undefined?

Author: Let me give a concrete example why I believe this _shouldn't_ be undefined behavior.

If we set this as undefined behavior it would be legal for the compiler to optimize the following shader to a no-op:

```hlsl
RWBuffer<float4> Buf;

[numthreads(64,1,1)]
void main(uint3 tid : SV_DispatchThreadID) {
    float4 Val = Buf[tid.x];
    float sum = Val.x + Val.y + Val.z + Val.w;
    Buf[tid.x].x = sum;
}
```
[Compiler Explorer](https://godbolt.org/z/n67s83Gxz)

That shader today compiles to both valid DXIL and valid SPIRV, and it does _something_ in each case. If we decide that a partial write to a typed buffer is undefined, the compiler would legally be allowed to eliminate the write, which would then mean it could legally eliminate the whole rest of the shader body.

That does not seem like the behavior we want.

Instead, if it is implementation defined, the compiler can't eliminate the write it is the responsibility of the implementation to have the partial write _do something_ and to document what that something is.

Author: I'm going to mark this conversation as resolved, because this language has been revised. As I commented [here](https://github.com/hlsl-tc57/tc57/pull/65/#discussion_r3533160073), @bogner and I investigated and found all implementations agree on this behavior so it can be specified and defined rather than implantation defined or undefined.

\Ch{Resource Types}{Resources}

\p Resource types are intangible types representing memory or configuration with an external interface.

\p These take the form of typed buffers, raw buffers, textures, constant buffers
and samplers.

\p Typed buffer, raw buffer, and Texture resources may be read-only or
read-write as denoted by the type of the resource, where the names of read-write
resource types are prefixed with \texttt{RW}.

\p Global resource declarations are \textit{bound} to device memory managed by
an application through a runtime implementation. A resource declaration that is
not bound is said to be \textit{unbound}. Operations on unbound resources are
evaluated as no-ops.

\p An implementation may impose restrictions on how local resource objects may
be used (\ref{AnnexB.LocalResources}).

\Sec{Typed Buffers}{Resources.TypedBuffers}

\p The typed buffer class template represents a one-dimensional resource
containing an array of a single given type. Its contents are indexed by typed
access to each element. Types may have their formats converted upon load in an
implementation-defined way.

\p The element type may be any type up to 16 bytes in size. Elements may be

> padded to 16 bytes if the element type is smaller than 16 bytes.

Reviewer 2: Not that it needs to be done here, but do we need to define padding rules/expected padding? If so, should we add a link to it here? Maybe it's too much detail, but anywhere `may` is used in a spec I have these thoughts.

Author: I agree the padding doesn't belong, and I think the constraint on the element type isn't quite accurate. I'm also noticing that we basically restate this constraint a couple paragraphs down. I'll rework this.

\p All typed buffers can be read through subscript operators or Load methods.
Writable typed buffers can be written through subscript operators.

\p Where \texttt{T} can be any arithmetic type \ref{Basic.Types.Arithmetic}

> or a vector with a maximum of four elements containing such an arithmetic type.

Reviewer 2: Annex B describes that an implementation may allow vectors of size greater than 4 elements. How does that interact with this clause?

The total size of a single contained element must be less than 16 bytes.

\p Typed buffers perform format conversions on load such that the underlying
data gets converted to the destination type in an implementation-defined way.

\p The \texttt{Buffer} type's template argument \texttt{T} has a default value
of \texttt{float4}.

\begin{HLSL}

> template <typename T = float4> requires hlsl::is_vector<T>::value || hlsl::is_arithmetic<T>::value

Reviewer 2: This line is rendering past the edge of the page in the pdf for me. Cuts off half way through `value` at the end.

class Buffer {
  Buffer();
  Buffer(const Buffer &buf);
  Buffer &operator=(Buffer &buf);
  Buffer(const __DynamicResourceHandle &handle);

  void GetDimensions(out uint width) const;

  // Element access.
  T operator[](int index) const;
  T Load(int index) const;
  T Load(int index, out uint status) const;
};

template <typename T> requires hlsl::is_vector<T>::value || hlsl::is_arithmetic<T>::value
class RWBuffer {
  RWBuffer();
  RWBuffer(RWBuffer buf);
  RWBuffer &operator=(RWBuffer &buf);

>   Buffer(const __DynamicResourceHandle &handle);

Reviewer 2: I'm not a fan of defining `__DynamicResourceHandle` here in the code listing then waiting a few points to define it in text. I think the text definition is fine, but I wish it were a little closer to the usage here.

  void GetDimensions(out uint width) const;

  // Element access.
  T operator[](int index) const;
  device T &operator[](int n) const;
  T Load(int index) const;
  T Load(int index, out uint status) const;
};
\end{HLSL}

\Sub{Constructors}{Resources.TypedBuffers.Ctrs}

\p When defined at global scope, typed buffers are bound to externally-defined
backing stores using the explicit binding location if provided or the implicit
binding if not (\ref{Resources.Binding}).

\p When defined at local scope, typed buffers represent local references that
can be associated with global buffers when assigned. Use of an uninitialized
typed buffer is undefined.

\begin{HLSL}
Buffer<int> grobuf;
RWBuffer<int> grwbuf;
void main() {
  Buffer<int> lrobuf = grobuf;

>   RWBuffer<int> lrwbuf = grwbuf;

Reviewer 2: It's not made explicitly clear what `lrobuf` `grobuf` `lrwbuf` and `grwbuf` are here, or if the names hold important meaning. I realise it maps to global/local read write/read only after thinking about it for a moment, but would be nicer if it were a little clearer.

}
\end{HLSL}

\p Buffers cannot be implicitly nor explicitly cast from one type to another.

\p Buffers may be created from dynamic resources. The
\texttt{\_\_DynamicResourceHandle} type in the examples above is a placeholder
for the implementation-specific type returned by subscript indexing into the
\texttt{ResourceDescriptorHeap} (\ref{Resources.Binding.Dynamic}).

\Sub{Dimensions}{Resources.TypedBuffers.Dims}

\p The size of a typed buffer can be retrieved using the \texttt{GetDimensions}
method.

\begin{HLSL}
void GetDimensions(out uint width) const;
\end{HLSL}

\p This returns the full length, in elements of the buffer through the

> \texttt{width} output parameter.

Reviewer 2: Curiosity rather than issue, but is there a reason we use an out param for `GetDimensions` rather than returning a `uint`. Is this just to keep uniformity with the texture variants of `GetDimensions`?

Author: I honestly have no idea when this decision was made, the spec is just matching the behavior. I suspect it is historical more than anything.

\Ch{Resource Types}{Resources}

> \p Resource types are intangible types representing memory or configuration with an external interface.

Reviewer 3: 80-col
```suggestion
\p Resource types are intangible types representing memory or configuration with
an external interface.
```

an application through a runtime implementation. A resource declaration that is
not bound is said to be \textit{unbound}. Operations on unbound resources are
evaluated as no-ops.

\p An implementation may impose restrictions on how local resource objects may
be used (\ref{AnnexB.LocalResources}).

\Sec{Optional Features}{Resources.Optional}

\p \textit{Dynamic resources} is an optional feature which allows resource
objects to be initialized by indexing directly into a heap of descriptors. A
conforming implementation must either support all interfaces related to dynamic
resources or none, but may not partially support them.

\p The \texttt{\_\_DynamicResourceHandle} type in the code examples in this
clause is a placeholder for the implementation-specific type returned by
subscript indexing into the \texttt{ResourceDescriptorHeap}
(\ref{Resources.Binding.Dynamic}).

\Sec{Special Semantics}{Resources.Special}

\Sub{Resource Initialization}{Resources.Special.Initialization}

\p A resource declaration in a global scope are called \textit{global
resource declaration} and represents a binding for external data into a
program. Global resource declarations are implicitly initialized in an

> implementation-defined.

Reviewer 3: ```suggestion
program. Global resource declarations are implicitly initialized in an
implementation-defined way.
```

read-write as denoted by the type of the resource, where the names of read-write
resource types are prefixed with \texttt{RW}.

\p Global resource declarations are \textit{bound} to device memory managed by
an application through a runtime implementation. A resource declaration that is
not bound is said to be \textit{unbound}. Operations on unbound resources are
evaluated as no-ops.

\p An implementation may impose restrictions on how local resource objects may
be used (\ref{AnnexB.LocalResources}).

\Sec{Optional Features}{Resources.Optional}

\p \textit{Dynamic resources} is an optional feature which allows resource
objects to be initialized by indexing directly into a heap of descriptors. A
conforming implementation must either support all interfaces related to dynamic
resources or none, but may not partially support them.

\p The \texttt{\_\_DynamicResourceHandle} type in the code examples in this
clause is a placeholder for the implementation-specific type returned by
subscript indexing into the \texttt{ResourceDescriptorHeap}
(\ref{Resources.Binding.Dynamic}).

\Sec{Special Semantics}{Resources.Special}

\Sub{Resource Initialization}{Resources.Special.Initialization}

> \p A resource declaration in a global scope are called \textit{global

Reviewer 2: ```suggestion
\p A resource declaration in a global scope is called a \textit{global
```

resource declaration} and represents a binding for external data into a
program. Global resource declarations are implicitly initialized in an
implementation-defined way.

\p Use of an uninitialized resource is undefined.

\Sec{Typed Buffers}{Resources.TypedBuffers}

\p The typed buffer class template represents a one-dimensional resource
containing an array of a single given type. Its contents are indexed by typed

> access to each element. The memory backing a resource need not formatted in the

Reviewer 2: ```suggestion
access to each element. The memory backing a resource need not be formatted in the
```

> same type as the resource's element type. If required, typed buffer elements are

Reviewer 4: This might just be my own mental model, but "formatted" seems odd here.  I'm accustomed to talking about memory layout in this kind of context ("The layout of the memory backing a resource need not match the layout of the resource's element type.").  If I'm alone here (or if I'm wrong or missed the point :-) ), I definitely don't feel strongly.

Author: The wording is definitely odd. The tough thing is that "layout" doesn't quite capture the nuance. The shader may specify that a typed buffer is comprised of 4 16-bit floating point values, but the underlying data could be 4 32-bit integers which are interpolated into 16-bit float values, so it can have both layout and type conversions (with interpolation rather than traditional conversions).

Reviewer 4: That's fair.  We can always update it if we find something more precise.

converted upon load in an implementation-defined way.

\p There are two typed buffer class templates \texttt{Buffer} and
\texttt{RWBuffer}, which respectively represent read-only and read-write typed
buffer objects.

\begin{HLSL}
template <typename T = float4>
requires hlsl::is_vector<T>::value || hlsl::is_arithmetic<T>::value
class Buffer {
  Buffer();
  Buffer(const Buffer &buf);
  Buffer &operator=(Buffer &buf);
  Buffer(const __DynamicResourceHandle &handle);

  void GetDimensions(out uint width) const;

  // Element access.
  T operator[](int index) const;
  T Load(int index) const;
  T Load(int index, out uint status) const;
};

template <typename T>
requires hlsl::is_vector<T>::value || hlsl::is_arithmetic<T>::value
class RWBuffer {
  RWBuffer();
  RWBuffer(RWBuffer buf);
  RWBuffer &operator=(RWBuffer &buf);
  Buffer(const __DynamicResourceHandle &handle);

  void GetDimensions(out uint width) const;

  // Element access.
  T operator[](int index) const;
  device T &operator[](int n) const;
  T Load(int index) const;
  T Load(int index, out uint status) const;
};
\end{HLSL}

\p The element type of a typed buffer, \texttt{T}, must be an arithmetic type

> (\ref{Basic.Types.Arithmetic}) or vector of arithmetic containing no more than 4

Reviewer 4: "...vector of arithmetic type..."

elements, and the total size of the element (\texttt{sizeof(T)}) must not exceed

> 16-bytes in size.

Reviewer 4: 16-bytes seems an odd limit.  Is this not meant to cover structured buffers?

Author: Yea, this only covers "typed buffers" which are the `Buffer` and `RWBuffer` objects. Structured buffers are "raw" buffers which have different rules.

Reviewer 4: Cool, thanks.

\end{itemize}

\begin{HLSL}
T Load(uint index) const;
\end{HLSL}

\begin{itemize}
  \item \textit{Requires:} \texttt{index} \textless the number of elements in the
  typed buffer.
  \item \textit{Returns:} The element in the typed buffer at the index specified
  by \texttt{index}.
\end{itemize}

\begin{HLSL}
T Load(int index, out uint status) const;
\end{HLSL}

\begin{itemize}
  \item \textit{Requires:} \texttt{index} \textless the number of elements in the
  typed buffer.
  \item \textit{Returns:} The element in the typed buffer at the index specified
  by \texttt{index}.
  \item \textit{Effects:} Initializes the \texttt{status} argument to an
  implementation-defined status signal (\ref{Resources.Mapcheck}).
  \item \textit{Remarks:} The \texttt{status} parameter returns an
  implementation-defined value that indicates if the loaded value came from
  fully mapped parts of the resource. This parameter must be passed to the
  built-in function \texttt{CheckAccessFullyMapped} (\ref{Resources.Mapcheck})
  in order to interpret it.
\end{itemize}

\Sub{Typed Buffer Operators}{Resources.TypedBuffers.Operators}

\begin{HLSL}
T operator[](uint index) const;
\end{HLSL}

\begin{itemize}

>   \item \textit{Requires:} \texttt{index} \textless the number of elements in the

Reviewer 2: I know you pushed a commit to fix the rendering from `<` to `\textless`, but I fear it's still rending incorrectly. I'm not seeing a space between  `<` and `the`. Same with the future uses of `\textless`. Very very minor.

  typed buffer.
  \item \textit{Returns:} The element in the typed buffer at the index specified
  by \texttt{index}.
\end{itemize}

\begin{HLSL}
device T &operator[](int index);
\end{HLSL}

\begin{itemize}

>   \item \textit{Preconditions:} \texttt{index} \textless the number of elements in the

Reviewer 2: For the previous operator you used `Requires` but here `Preconditions` for what is the same requirement. I believe both should be `Requires`

```suggestion
  \item \textit{Requires:} \texttt{index} \textless the number of elements in the
```

  typed buffer.
  \item \textit{Returns:} An lvalue-reference to the element in the typed buffer
  at the index specified by \texttt{index}.
  \item \textit{Remarks:} Writing to typed buffer elements overwrite the entire
  element, even if only part of the data is written. A partial write to a typed
  buffer is a load-insert-store operation.
\end{itemize}

\begin{note}
  Accurately modeling the memory operations for writable typed buffers would
  require a proxy-object interface since the address written to isn't actually
  the device memory since it may be converted or otherwise normalized by the
  implementation.
\end{note}

\Sec{CheckAccessFullyMapped}{Resources.Mapcheck}

\p The implementation-defined status value returned by texture and buffer load
methods is interpreted by a built-in function:

\begin{HLSL}
bool CheckAccessFullyMapped(uint status);
\end{HLSL}

\begin{itemize}

>   \item \textit{Preconditions:} \texttt{status} is an initialized value from a

Reviewer 2: ```suggestion
  \item \textit{Requires:} \texttt{status} is an initialized value from a
```
Same as above.

\p An implementation may impose restrictions on how local resource objects may
be used (\ref{AnnexB.LocalResources}).

\Sec{Optional Features}{Resources.Optional}

\p \textit{Dynamic resources} is an optional feature which allows resource
objects to be initialized by indexing directly into a heap of descriptors. A
conforming implementation must either support all interfaces related to dynamic
resources or none, but may not partially support them.

\p The \texttt{\_\_DynamicResourceHandle} type in the code examples in this
clause is a placeholder for the implementation-specific type returned by
subscript indexing into the \texttt{ResourceDescriptorHeap}

> (\ref{Resources.Binding.Dynamic}).

Reviewer: This section doesn't exist and was missing from hlsl-specs as well.

Author: In many other PRs I've created references to sections that don't yet exist. LaTeX provides nice diagnostics of these that are easy to track. If this PR is approved by the committee I'll file an issue for the unwritten section to track populating it.

> \Sec{Special Semantics}{Resources.Special}

Reviewer: Oh I don't like that the subsection has to be called this. It doesn't reveal much of what the content is.

Author: Do you have an alternative suggestion?

\Sub{Resource Initialization}{Resources.Special.Initialization}

\p A resource declaration in a global scope are called \textit{global
resource declaration} and represents a binding for external data into a

> program. Global resource declarations are implicitly initialized in an

Reviewer: Is the intent to add the Resources.Binding section eventually or just leave it at "implementation-defined way"?

Author: I expect to add a section that describes the resource binding attributes, but not the specific semantics of binding since they vary by implementation, so this statement will remain valid.

\p A resource declaration in a global scope are called \textit{global
resource declaration} and represents a binding for external data into a
program. Global resource declarations are implicitly initialized in an
implementation-defined way.

\p The element type may be any type up to 16 bytes in size. Elements may be
padded to 16 bytes if the element type is smaller than 16 bytes.
\p Use of an uninitialized resource is undefined.

\p All typed buffers can be read through subscript operators or Load methods.
Writable typed buffers can be written through subscript operators.
\Sec{Typed Buffers}{Resources.TypedBuffers}

\p Where \texttt{T} can be any arithmetic type \ref{Basic.Types.Arithmetic}
or a vector with a maximum of four elements containing such an arithmetic type.

> The total size of a single contained element must be less than 16 bytes.

Reviewer: Why remove this information? At the least, mention that a type exceeding a certain size or number of elements may be disallowed by an implementation

Author: Given the extent of the churn on this, it is probably easier to review the final state of the diff rather than individual commits. This information is still present in the specification, it just isn't at this location in the file. It is [here](https://github.com/hlsl-tc57/tc57/pull/65/changes#diff-f43f5ddb995c0acc6ae11cd5c41b19f95a84457e4416d8fe0405b3926e7df1c7R95)

  \item \textit{Remarks:} The \texttt{status} parameter returns an
  implementation-defined value that indicates if the loaded value came from
  fully mapped parts of the resource. This parameter must be passed to the
  built-in function \texttt{CheckAccessFullyMapped} (\ref{Resources.Mapcheck})
  in order to interpret it.
\end{itemize}

\Sub{Typed Buffer Operators}{Resources.TypedBuffers.Operators}

\begin{HLSL}
T operator[](uint index) const;
\end{HLSL}

\begin{itemize}
  \item \textit{Requires:} \texttt{index} \textless the number of elements in the
  typed buffer.
  \item \textit{Returns:} The element in the typed buffer at the index specified
  by \texttt{index}.
\end{itemize}

\begin{HLSL}
device T &operator[](int index);
\end{HLSL}

\begin{itemize}
  \item \textit{Preconditions:} \texttt{index} \textless the number of elements in the
  typed buffer.
  \item \textit{Returns:} An lvalue-reference to the element in the typed buffer
  at the index specified by \texttt{index}.
  \item \textit{Remarks:} Writing to typed buffer elements overwrite the entire
  element, even if only part of the data is written. A partial write to a typed

>   buffer is a load-insert-store operation.

Reviewer: Why the change from "implementation-defined"? Where is "load-insert-store" described?

Author: @bogner and I investigated and found that DXC and Clang for both DirectX and Vulkan behave consistently so we decided to codify that behavior.

\end{itemize}

\begin{note}
  Accurately modeling the memory operations for writable typed buffers would
  require a proxy-object interface since the address written to isn't actually
  the device memory since it may be converted or otherwise normalized by the
  implementation.
\end{note}

\Sec{CheckAccessFullyMapped}{Resources.Mapcheck}

\p The implementation-defined status value returned by texture and buffer load

> methods is interpreted by a built-in function:

Reviewer: Why was the description of the status parameter altered only here to merely be identified as "implementation-defined"?

Author: It is identified [here](https://github.com/hlsl-tc57/tc57/pull/65/changes#diff-f43f5ddb995c0acc6ae11cd5c41b19f95a84457e4416d8fe0405b3926e7df1c7R176) also as being implementation-defined. Are there other places that you think it should be clarified that its value is implementation defined?
